storage/localstore/mode_set.go

```diff
@@ -201,48 +201,65 @@ func (db *DB) setSync(batch *leveldb.Batch, addr chunk.Address, mode chunk.ModeS
 	item.StoreTimestamp = i.StoreTimestamp
 	item.BinID = i.BinID
 
-	// moveToGc toggles the deletion of the item from pushsync index
-	// it will be false in the case pull sync was called but push sync was meant on that chunk (!tag.Anonymous)
-	moveToGc := true
-	if db.tags != nil {
-		i, err = db.pushIndex.Get(item)
-		switch err {
-		case nil:
-			tag, _ := db.tags.Get(i.Tag)
-			if tag != nil {
-				if tag.Anonymous {
-					// anonymous - only pull sync
-					switch mode {
-					case chunk.ModeSetSyncPull:
-						// this will not get called twice because we remove the item once after the !moveToGc check
-						tag.Inc(chunk.StateSent)
-						// this is needed since pushsync is checking if `tag.Done(chunk.StateSynced)` and when overlapping
-						// chunks are synced by both push and pull sync we have a problem. as an interim solution we increment this too
-						tag.Inc(chunk.StateSynced)
-					case chunk.ModeSetSyncPush:
-						// do nothing - this should not be possible. just log and return
-						log.Warn("chunk marked as push-synced but should be anonymous!", "tag.Uid", tag.Uid)
-						moveToGc = false
-					}
-				} else {
-					// not anonymous - push and pull should be used
-					switch mode {
-					case chunk.ModeSetSyncPull:
-						moveToGc = false
-					case chunk.ModeSetSyncPush:
-						tag.Inc(chunk.StateSynced)
-					}
-				}
+	switch mode {
+	case chunk.ModeSetSyncPull:
+		// if we are setting a chunk for pullsync we expect it to be in the index
+		// if it has a tag - we increment it and set the index item to _not_ contain the tag reference
+		// this prevents duplicate increments
+		i, err := db.pullIndex.Get(item)
+		if err != nil {
+			if err == leveldb.ErrNotFound {
+				// we handle this error internally, since this is an internal inconsistency of the indices
+				// if we return the error here - it means that for example, in stream protocol peers which we sync
+				// to would be dropped. this is possible when the chunk is put with ModePutRequest and ModeSetSyncPull is
+				// called on the same chunk (which should not happen)
+				log.Error("chunk not found in pull index", "addr", addr)
+				break
 			}
+			return 0, err
+		}
 
-		case leveldb.ErrNotFound:
-			// the chunk is not accessed before
-		default:
+		if db.tags != nil && i.Tag != 0 {
+			t, err := db.tags.Get(i.Tag)
+
+			// increment if and only if tag is anonymous
+			if err == nil && t.Anonymous {
+				t.Inc(chunk.StateSent)
+				t.Inc(chunk.StateSynced)
+				item.Tag = 0
```

Comment:
This is super important, this line makes sure that syncing to several peers will not increment the same tag twice. 
Note that in push sync we delete synced chunks so we never increment twice. 
Please add a comment that explains this